### PR TITLE
Deploy charmhub bundle with charmhub charms

### PR DESCRIPTION
**Pull request by SimonRichardson**
When attempting to resolve an origin we should populate all the fields
of the origin so that we can correctly send them to and from the server.
These items were missed because of the changes to the architecture, but
have now been correctly filled in.

Deploying a local charmhub charm is now possible thanks to this.

## QA steps

```sh
$ export JUJU_DEV_FEATURE_FLAGS=charm-hub
$ juju bootstrap lxd test --config charm-hub-url="https://api.staging.snapcraft.io"
$ juju deploy ./testcharms/charm-hub/bundles/wordpress-simple/bundle.yaml
Resolving charm via charmhub: ch:mysql
Resolving charm via charmhub: ch:wordpress
Executing changes:
- upload charm ch:wordpress-9
- deploy application wordpress using ch:wordpress-9
ERROR cannot deploy bundle: cannot deploy application "wordpress": cannot add application "wordpress": series "kubernetes" in a non container model not valid
```

Ensure that mysql got installed correctly.

```sh
$ juju status
Model    Controller  Cloud/Region  Version    SLA          Timestamp
default  test        lxd/default   2.9-rc3.1  unsupported  15:54:57Z

App    Version  Status   Scale  Charm  Store     Rev  OS      Message
mysql           unknown      0  mysql  charmhub   58  ubuntu
```

### CODE REVIEW

File: apiserver/facades/client/charms/client.go [hunk 1]
```diff
@@ -521,7 +521,24 @@ func (a *API) resolveOneCharm(arg params.ResolveCharmWithChannel, mac *macaroon.
 		return result
 	}
 	result.URL = resultURL.String()
-	result.Origin = origin
+
+	// We have an issue where the charmhub API can return "all" for architecture
```
Comment (manadart) [inline apiserver/facades/client/charms/client.go:525]: We have an issue that can cause us issues? :stuck_out_tongue:

File: apiserver/facades/client/charms/client.go [hunk 2]
```diff
@@ -521,7 +521,24 @@ func (a *API) resolveOneCharm(arg params.ResolveCharmWithChannel, mac *macaroon.
...
+	// and this can cause us issues. To try and solve that, we will use the
+	// model constraints to resolve it.
+	archOrigin := origin
+	if origin.Architecture == "all" {
+		cons, err := a.backendState.ModelConstraints()
+		if err != nil {
+			result.Error = apiservererrors.ServerError(err)
+			return result
+		}
+		if cons.HasArch() {
```
Comment (manadart) [inline apiserver/facades/client/charms/client.go:535]: If there is no model constraint, we still get "all" though, and have aforementioned issues?
Not common to have such a constraint AFAIK.

File: apiserver/facades/client/charms/repositories.go [hunk 3]
```diff
@@ -58,7 +58,13 @@ func (c *chRepo) ResolveWithPreferredChannel(curl *charm.URL, origin params.Char
 	// If no revision nor channel specified, use the default release.
 	if curl.Revision == -1 && channel.String() == "" {
 		logger.Debugf("Resolving charm with default release")
-		return c.resolveViaChannelMap(info.Type, curl, origin, info.DefaultRelease)
+		resURL, resOrigin, series, err := c.resolveViaChannelMap(info.Type, curl, origin, info.DefaultRelease)
```
Comment (manadart) [inline apiserver/facades/client/charms/repositories.go:61]: _series_ collides with an import in this file.

**Review by manadart — APPROVED** [review]
QA is good. Couple of suggestions. Don't know how to make the "all" situation watertight.

File: apiserver/facades/client/charms/client.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/facades/client/charms/client.go:525]: Yes, lots of issues

File: apiserver/facades/client/charms/client.go (hunk 2 above)
Comment (SimonRichardson) [reply apiserver/facades/client/charms/client.go:535]: I'm speaking to the charmhub guys to resolve this.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**